design/storage-proof-timing.md

```diff
@@ -0,0 +1,173 @@
+leading to more transactions, leading to even more blocks, etc.
+
+And finally, because storage contracts between a client and a host are best
+expressed in wall clock time, there are going to be two different ways of
+measuring time in the same smart contract, which could lead to some subtle bugs.
+
+These problems lead us to the second design.
+
+Design 2: block pointers
+------------------------
+
+In our second design we separate cadence from random number selection. For
+cadence we use a time interval measured in seconds. This divides time into
+periods that have a unique number. Each period represents a chance that a proof
+is required.
+
+We want to associate a random number with each period that is used for the proof
+of storage challenge, and for the die roll to determine whether a proof is
+required. But since we no longer have a one-on-one relationship between a period
+and a block hash, we need to get creative.
+
+EVM and solidity
+----------------
+
+For context, our smart contracts are written in Solidity and execute on the EVM.
+In this environment we have access to the [most recent 256 block hashes][1], and
+to the current time, but not to the timestamps of the previous blocks. We also
+have access to the current block number.
+
+Block pointers
+--------------
+
+We introduce the notion of a block pointer. This is a number between 0 and 256
+that points to one of the latest 256 block hashes. We count from 0 (latest
+block) to 255 (oldest available block).
+
+         oldest                              latest
+     - - - |-----------------------------------|
+          255                    ^             0
+                                 |
+                              pointer
+
+We want to associate a block pointer with a period such that it keeps pointing
+to the same block hash when new blocks are produced. We need this because the
+block hash is used to check whether a proof is required, to check a proof when
+it's submitted, or to prove absence of a proof, all at different times.
+
+To ensure that the block pointer points to the same block hash for longer
+periods of time, we derive it from the current block number:
+
+    pointer(period) = (blocknumber + period) % 256
+
+Over time, when more blocks are produced, we get this picture:
+
+     |
+     |      - - - |-----------------------------------|
+     |           255                          ^       0
+     |                                        |
+     |                                     pointer
+     t
+     i      - - - |-----------------------------------|
+     m           255                      ^           0
+     e                                    |
+     |                                  pointer
+     |
+     |      - - - |-----------------------------------|
+     |           255                  ^               0
+     |                                |
+     v                             pointer
+
+Pointer duos
```

Comment:
What if we simply use a smaller modulo, wouldn't that give us a window that will keep the pointer valid for that many blocks - `WINDOW = 256 - (blocknumber + period) % modulo`? For example, if the modulo is 128, then this gives the node a `WINDOW` of at least 128 blocks where the pointer is still valid. Once the `WINDOW` has expired the proof is impossible to verify? We can adjust the `WINDOW` depending on the average block production cadence or some other parameter, even wall clock?

Reply:
Obviously we adjust the window by tweaking the `modulo`.